Add EP249 summary blog post (#2673)
quarkusio/quarkusio.github.io: _posts/2026-06-01-quarkus-insights-249-coding-agents.adoc

tags: insights ai coding-agents mcp claude developer-experience
synopsis: 'Philip Krüger demonstrates the Quarkus Agent MCP server, showing how coding agents can enhance the developer experience through skills, tools, and documentation integration for building Quarkus applications.'
author: jcobb
---

This summary was generated using AI, reviewed by humans - https://youtube.com/live/EczihvU1gG8[watch the video] for the full story.

In episode 249 of Quarkus Insights, Philip Krüger from Australia joined the team to demonstrate how coding agents can enhance the https://quarkus.io/[Quarkus] developer experience. Despite battling flu season and joining at 11 PM his time, Philip showcased the new https://github.com/quarkusio/quarkus-agent-mcp[Quarkus Agent MCP] server and how it enables AI-powered development workflows.

++++
<iframe width="560" height="315" src="https://www.youtube.com/embed/EczihvU1gG8" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>
++++

== News: Quarkus 3.36 Platform Release

> Before diving into coding agents, the team announced the https://quarkus.io/blog/[Quarkus 3.36 platform release], which includes:

Reviewer: ```suggestion
Before diving into coding agents, the team announced the https://quarkus.io/blog/quarkus-3-36-released/[Quarkus 3.36 platform release], which includes:
```

> - **Tree shaking** for optimized builds

Reviewer: I said this, but I was wrong - it was in 3.35. Let's add an arbitrary other feature instead so the bullets look ok. 
```suggestion
- **OIDC Spiffe authentication
```

- **Signals** - a new eventing infrastructure
- Enhanced developer experience features

> For those interested in the design decisions behind signals, the team recorded a detailed presentation during the Quarkus community call with extensive discussion about the architecture and future direction.

Reviewer: ```suggestion
For those interested in the design decisions behind signals, the team recorded a https://www.youtube.com/watch?v=_VY9G_DMFjw[detailed presentation] during the Quarkus community call with extensive discussion about the architecture and future direction.
```

- **For agents**: Explicit reload command when ready

=== 2. Add Extension Skills

Create a skill file in your extension:

[source,text]
----
deployment/src/main/resources/quarkus-skill.md
----

The skill should include:
- Clear description of the extension's purpose
- Usage patterns and examples
- Common pitfalls and solutions
- Integration points with other extensions

=== 3. Ensure Documentation Quality

Make sure your extension's documentation is:
- Accurate and up-to-date
- Well-structured with clear headings
- Includes practical examples
- Links to related documentation

For core Quarkus extensions, documentation is already integrated. For https://github.com/quarkiverse[Quarkiverse] extensions, add this to your POM:

[source,xml]
----
<plugin>
  <groupId>io.quarkus</groupId>
  <artifactId>quarkus-doc-maven-plugin</artifactId>
  <version>${quarkus.version}</version>
</plugin>
----

=== 4. Follow the Extension Maturity Matrix

The https://quarkus.io/guides/extension-maturity-matrix[Extension Maturity Matrix] now includes AI experience criteria. Following these guidelines ensures your extension works well with coding agents.

== Team's AI Development Practices

> The team shared their diverse approaches to AI-assisted development:

Reviewer: ```suggestion
In response to questions on the live-stream chat, the team shared their diverse approaches to AI-assisted development:
```

=== Tools and Models Used

**IDEs and Harnesses**:
- https://claude.ai/[Claude Desktop] (most popular)

> - https://www.jetbrains.com/idea/[IntelliJ IDEA] with AI Assistant

Reviewer: ```suggestion
- https://www.jetbrains.com/idea/[IntelliJ IDEA] with AI Assistant
- https://bob.ibm.com/
```

- https://www.cursor.com/[Cursor]
- https://github.com/Pythagora-io/gpt-pilot[GPT Pilot]
- https://www.jetbrains.com/qodana/[JetBrains Qodana]

**Models**:
- Claude (Sonnet and Opus)
- https://github.com/features/copilot[GitHub Copilot]
- https://ai.google.dev/gemini-api[Google Gemini]

> - https://www.ibm.com/watsonx[IBM watsonx]

Reviewer: I don't think we mentioned this, so we could drop it? To balance the IBM addition in the previous list? No strong feelings!

Author: Removed line 350 regarding WatsonX

=== Workflow Patterns

**Philip's Approach**:
- Uses Claude Desktop exclusively
- Views code through Dev UI's workspace browser
- Treats the agent as a soundboard for design discussions
- Runs multiple Claude instances for different tasks
- Rarely opens a traditional IDE

**Max's Approach**:
- Prefers https://github.com/Pythagora-io/gpt-pilot[GPT Pilot]
- Focuses on incremental development
- Uses plan mode but avoids over-planning
- Emphasizes iterative refinement

**Eric's Approach**:
- Uses IntelliJ with Claude integration
- Leverages IDE context sensitivity
- Benefits from IDE's diff viewer
- Combines IDE refactoring with AI generation

**Holly's Approach**:

> - Swaps between Claude and Qodana

Reviewer: ```suggestion
- Swaps between Claude Code and Bob
```

- **Dependency updates**: Handling breaking changes
- **CI/CD maintenance**: Fixing build failures
- **Boilerplate generation**: CRUD operations, DTOs, mappers
- **Documentation**: Generating or updating docs
- **Refactoring**: Mechanical code transformations

=== Less Effective Use Cases

- **Complex architecture decisions**: Still requires human judgment
- **Novel algorithms**: Agents struggle with truly new approaches
- **Performance optimization**: Requires deep understanding
- **Security-critical code**: Needs expert review

=== The "Soundboard" Pattern

Philip highlighted using agents as a soundboard for design discussions, especially valuable when working across time zones without human colleagues available.

== Challenges and Future Directions

=== Current Challenges

**Non-deterministic behavior**: Same prompt can produce different results, making testing difficult.

**Rapid evolution**: The AI landscape changes quickly, requiring constant adaptation.

**Tool proliferation**: Managing the growing ecosystem of AI tools and approaches.

**Context management**: Balancing between providing enough context and overwhelming the model.

=== Future Improvements

> **Better sandboxing**: The team mentioned https://get-inca.com/en/[Inca Sandbox] for safer agent execution.

Reviewer: ```suggestion
**Better sandboxing**: The team mentioned https://github.com/Sanne/incus-spawn[Incus Spawn] for safer agent execution.
```
